Rewrite ASM module

Monkeyanator opened the pull request:
The existing ASM module is a near-1:1 wrapper around `install_asm` which is not ideal. This rewritten module uses the ControlPlaneRevision KRM API to provision the control plane. We still breaks out and use a provisioner to apply the CPR (due to limitations around the `kubernetes_manifest` resource) but it's much slimmer than before, and doesn't involve curling down an installation script.

comment-bot-dev commented:
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_shell
 Failed shell check. More info on running shellcheck locally [here](https://www.shellcheck.net). 
 ```bash 
Running shellcheck
In ./modules/asm/scripts/destroy_cpr.sh line 26:
if !kubectl delete controlplanerevision -n istio-system ${REVISION_NAME} ; then
    ^-- SC1035: You are missing a required space after the !.
                                                        ^--------------^ SC2086: Double quote to prevent globbing and word splitting.
In ./modules/asm/scripts/create_cpr.sh line 26:
for i in {1..6}; do kubectl get crd ${CPR_RESOURCE} && break || sleep 10; done
^-^ SC2034: i appears unused. Verify use (or export if used externally).
                                                    ^-- SC2015: Note that A && B || C is not if-then-else. C may run when A is true.
In ./modules/asm/scripts/create_cpr.sh line 46:
kubectl wait -n istio-system --for=condition=Reconciled controlplanerevision/${REVISION_NAME} --timeout 5m
                                                                             ^--------------^ SC2086: Double quote to prevent globbing and word splitting.
For more information:
  https://www.shellcheck.net/wiki/SC1035 -- You are missing a required space ...
  https://www.shellcheck.net/wiki/SC2034 -- i appears unused. Verify use (or ...
  https://www.shellcheck.net/wiki/SC2015 -- Note that A && B || C is not if-t...
 ``` 
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
examples/simple_zonal_with_asm/main.tf
--- old/examples/simple_zonal_with_asm/main.tf
+++ new/examples/simple_zonal_with_asm/main.tf
@@ -57,9 +57,9 @@
 }
 
 module "asm" {
-  source                    = "../../modules/asm"
-  cluster_name              = module.gke.name
-  cluster_location          = module.gke.location
-  project_id                = var.project_id
-  location                  = module.gke.location
+  source           = "../../modules/asm"
+  cluster_name     = module.gke.name
+  cluster_location = module.gke.location
+  project_id       = var.project_id
+  location         = module.gke.location
 }
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
examples/simple_zonal_with_asm/hub.tf
--- old/examples/simple_zonal_with_asm/hub.tf
+++ new/examples/simple_zonal_with_asm/hub.tf
@@ -15,10 +15,10 @@
  */
 
 resource "google_gke_hub_membership" "cluster_membership" {
-  provider = google-beta
-  project = var.project_id
+  provider      = google-beta
+  project       = var.project_id
   membership_id = "${module.gke.name}-membership"
-  endpoint{
+  endpoint {
     gke_cluster {
       resource_link = "//container.googleapis.com/${module.gke.cluster_id}"
     }
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
modules/asm/main.tf
--- old/modules/asm/main.tf
+++ new/modules/asm/main.tf
@@ -16,19 +16,19 @@
 
 locals {
   // GKE release channel is a list with max length 1 https://github.com/hashicorp/terraform-provider-google/blob/9d5f69f9f0f74f1a8245f1a52dd6cffb572bbce4/google/resource_container_cluster.go#L954
-  gke_release_channel = length(data.google_container_cluster.asm_cluster.release_channel) > 0 ? data.google_container_cluster.asm_cluster.release_channel[0].channel : ""
+  gke_release_channel       = length(data.google_container_cluster.asm_cluster.release_channel) > 0 ? data.google_container_cluster.asm_cluster.release_channel[0].channel : ""
   gke_release_channel_fixed = local.gke_release_channel == "UNSPECIFIED" ? "" : local.gke_release_channel
   // In order or precedence, use (1) user specified channel, (2) GKE release channel, and (3) regular channel
-  channel = lower(coalesce(var.channel, local.gke_release_channel_fixed, "regular"))
-  revision_name = "asm-managed${local.channel == "regular" ? "" : "-${local.channel}"}"
+  channel          = lower(coalesce(var.channel, local.gke_release_channel_fixed, "regular"))
+  revision_name    = "asm-managed${local.channel == "regular" ? "" : "-${local.channel}"}"
   mesh_config_name = "istio-${local.revision_name}"
   // CNI should be enabled if either enable_cni or enable_mdp are set
   enable_cni = var.enable_cni || var.enable_mdp
 }
 
 data "google_container_cluster" "asm_cluster" {
-  project = var.project_id
-  name = var.cluster_name
+  project  = var.project_id
+  name     = var.cluster_name
   location = var.cluster_location
 
   // This evaluates during planning phase unless we explicitly require a dependency on
@@ -44,21 +44,21 @@
 }
 
 resource "kubernetes_config_map" "mesh_config" {
-    metadata {
-      name = local.mesh_config_name
-      namespace = kubernetes_namespace.system_namespace.metadata[0].name
-      labels = {
-        "istio.io/rev" = local.revision_name
-      }
-    }
-    data = {
-      mesh = yamlencode(var.mesh_config)
+  metadata {
+    name      = local.mesh_config_name
+    namespace = kubernetes_namespace.system_namespace.metadata[0].name
+    labels = {
+      "istio.io/rev" = local.revision_name
     }
+  }
+  data = {
+    mesh = yamlencode(var.mesh_config)
+  }
 }
 
 resource "kubernetes_config_map" "asm_options" {
   metadata {
-    name = "asm-options"
+    name      = "asm-options"
     namespace = kubernetes_namespace.system_namespace.metadata[0].name
   }
 
@@ -68,18 +68,18 @@
 }
 
 resource "google_gke_hub_feature" "mesh_feature" {
-  name = "servicemesh"
-  project = var.project_id
+  name     = "servicemesh"
+  project  = var.project_id
   location = "global"
   provider = google-beta
 }
 
 module "cpr" {
-  source  = "terraform-google-modules/gcloud/google//modules/kubectl-wrapper"
+  source = "terraform-google-modules/gcloud/google//modules/kubectl-wrapper"
 
-  project_id                  = var.project_id
-  cluster_name                = var.cluster_name
-  cluster_location            = var.cluster_location
+  project_id       = var.project_id
+  cluster_name     = var.cluster_name
+  cluster_location = var.cluster_location
 
   kubectl_create_command  = "${path.module}/scripts/create_cpr.sh ${local.revision_name} ${local.channel} ${local.enable_cni}"
   kubectl_destroy_command = "${path.module}/scripts/destroy_cpr.sh ${local.revision_name}"
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
modules/asm/variables.tf
--- old/modules/asm/variables.tf
+++ new/modules/asm/variables.tf
@@ -16,22 +16,22 @@
 
 variable "project_id" {
   description = "The project in which the resource belongs."
-  type = string
+  type        = string
 }
 
 variable "cluster_name" {
   description = "The unique name to identify the cluster in ASM."
-  type = string
+  type        = string
 }
 
 variable "cluster_location" {
   description = "The cluster location for this ASM installation."
-  type = string
+  type        = string
 }
 
 variable "channel" {
   description = "The channel to use for this ASM installation."
-  type = string
+  type        = string
   validation {
     condition = anytrue([
       var.channel == "rapid",
@@ -46,26 +46,26 @@
 
 variable "enable_cni" {
   description = "Determines whether to enable CNI for this ASM installation."
-  type = bool
-  default = true
+  type        = bool
+  default     = true
 }
 
 // This should be validated so that it cannot be enabled while CNI is disabled
 // but validating based on other variables is not possible today (https://github.com/hashicorp/terraform/issues/25609)
 variable "enable_mdp" {
   description = "Determines whether to enable Managed Data Plane (MDP) for this ASM installation."
-  type = bool
-  default = true
+  type        = bool
+  default     = true
 }
 
 variable "enable_cross_cluster_service_discovery" {
   description = "Determines whether to enable cross-cluster service discovery between this cluster and other clusters in the fleet."
-  type = bool
-  default = false
+  type        = bool
+  default     = false
 }
 
 variable "mesh_config" {
   description = "MeshConfig specifies configuration available to the control plane. The full list of options can be found at https://istio.io/latest/docs/reference/config/istio.mesh.v1alpha1/#MeshConfig"
-  type = map
-  default = {}
+  type        = map(any)
+  default     = {}
 }
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
modules/asm/outputs.tf
--- old/modules/asm/outputs.tf
+++ new/modules/asm/outputs.tf
@@ -15,6 +15,6 @@
  */
 
 output "revision_name" {
-  value = local.revision_name
+  value       = local.revision_name
   description = "The name of the installed managed ASM revision."
 }
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
 ```

morgante commented on examples/simple_zonal_with_asm/hub.tf:
```diff
@@ -14,11 +14,21 @@
-  cluster_name            = module.gke.name
-  cluster_endpoint        = module.gke.endpoint
-  gke_hub_membership_name = "gke-asm-membership"
+resource "google_gke_hub_feature" "mesh_feature" {
```
This should probably be part of the module.

Monkeyanator replied:
Agreed, blocked on the provider bug described below.

morgante commented on modules/asm/main.tf:
```diff
@@ -14,95 +14,76 @@
-  kubectl_create_command_base = "${path.module}/scripts/install_asm.sh ${var.project_id} ${var.cluster_name} ${var.location} ${var.asm_version} ${var.mode} ${var.managed_control_plane} ${var.skip_validation} ${local.options_string} ${local.custom_overlays_string} ${var.enable_all} ${var.enable_cluster_roles} ${var.enable_cluster_labels} ${var.enable_gcp_components} ${var.enable_registration} ${va ...
+  // GKE release channel is a list with max length 1 https://github.com/hashicorp/terraform-provider-google/blob/9d5f69f9f0f74f1a8245f1a52dd6cffb572bbce4/google/resource_container_cluster.go#L954
+  gke_release_channel       = data.google_container_cluster.asm_cluster.release_channel != null ? data.google_container_cluster.asm_cluster.release_channel[0].channel : ""
+  gke_release_channel_fixed = local.gke_release_channel == "UNSPECIFIED" ? "" : local.gke_release_channel
```
Should probably just collapse this into the line above or below.

Monkeyanator replied:
Done.

morgante commented on modules/asm/main.tf:
```diff
@@ -14,95 +14,76 @@
-module "asm-services" {
-  source  = "terraform-google-modules/project-factory/google//modules/project_services"
-  version = "~> 11.3"
+resource "kubernetes_namespace" "system_namespace" {
```
```suggestion
resource "kubernetes_namespace" "system" {
```

Monkeyanator replied:
Done.

morgante commented on modules/asm/main.tf:
```diff
@@ -14,95 +14,76 @@
 
-resource "google_project_iam_member" "asm_iam" {
-  for_each = toset(local.asm_iam_member != "" ? (var.managed_control_plane ? local.mcp_iam_roles : local.iam_roles) : [])
+data "google_container_cluster" "asm_cluster" {
```
`asm` or `main`. `_cluster` is redundant (as with all such names).

Monkeyanator replied:
Fixed, thanks.

morgante commented on modules/asm/main.tf:
```diff
@@ -14,95 +14,76 @@
-  disable_services_on_destroy = false
-  disable_dependent_services  = false
+  metadata {
+    name      = local.mesh_config_name
```
This is only used once, so inline it. (The indirection through a local isn't helpful.)

Monkeyanator replied:
Done.

morgante commented on modules/asm/main.tf:
```diff
@@ -14,95 +14,76 @@
+  }
+}
+
+# TODO(Monkeyanator) due to a bug in the gke_hub_feature resource implementation this fails when enabling an enabled
```
Is there an upstream bug filed?

Monkeyanator replied:
Not yet, will update here once created.

Monkeyanator replied:
I'm unable to reproduce the bug, added the feature enablement back here.

morgante commented on modules/asm/scripts/create_cpr.sh:
```diff
@@ -0,0 +1,55 @@
+    exit 1
+fi
+
+# Wait for the CRD to get created before creating the CPR.
```
It would be good to explain why we need to do this instead of `kubectl --wait`.

Monkeyanator replied:
Done.

morgante commented on modules/asm/scripts/create_cpr.sh:
```diff
@@ -0,0 +1,55 @@
+
+kubectl wait --for condition=established --timeout=60s crd/"${CPR_RESOURCE}"
+
+REVISION_NAME=$1; shift
```
These should be at the top of the script.

Monkeyanator replied:
Done.

morgante commented on test/integration/simple_zonal_with_asm/controls/kubectl.rb:
```diff
@@ -41,18 +41,5 @@
       )
     end
-
-    describe "Mesh" do
```
Please don't remove the tests. If this needs to be updated to work with newer ASM, please adapt it but we want to continue verifying ASM was actually installed.

Monkeyanator replied:
Yeah I'm going to flesh out the tests to check for CPR existence. This one had to be removed since we don't install gateways in the new module.

Monkeyanator replied:
Expanded the tests.

kaariger commented on modules/asm/main.tf:
```diff
@@ -14,95 +14,57 @@
-  project_id = var.project_id
+locals {
+  // GKE release channel is a list with max length 1 https://github.com/hashicorp/terraform-provider-google/blob/9d5f69f9f0f74f1a8245f1a52dd6cffb572bbce4/google/resource_container_cluster.go#L954
+  gke_release_channel = data.google_container_cluster.asm.release_channel != null ? data.google_container_cluster.asm.release_channel[0].channel : ""
```
This does not handle the case where the GKE module returns "unspecified" as the release channel. It results in the same string being passed to the cpr module.

Monkeyanator replied:
Fixed, thanks.

Monkeyanator commented on modules/asm/outputs.tf:
```diff
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-output "asm_wait" {
```
Added it back, thanks.

Monkeyanator commented on modules/asm/main.tf:
```diff
@@ -14,95 +14,57 @@
+#}
+
+module "cpr" {
+  source = "terraform-google-modules/gcloud/google//modules/kubectl-wrapper"
```
Done.

Monkeyanator commented:
Addressed comments. CI failing on an issue that doesn't seem related to these changes though, any thoughts here? @bharathkkb 

```
Step #27 - "converge simple-zonal-private-local":        
Step #27 - "converge simple-zonal-private-local":        Error: No value for required variable
Step #27 - "converge simple-zonal-private-local":        
Step #27 - "converge simple-zonal-private-local":          on variables.tf line 17:
Step #27 - "converge simple-zonal-private-local":          17: variable "project_ids" {
Step #27 - "converge simple-zonal-private-local":        
Step #27 - "converge simple-zonal-private-local":        The root module input variable "project_ids" is not set, and has no default
Step #27 - "converge simple-zonal-private-local":        value. Use a -var or -var-file command line argument to provide a value for
Step #27 - "converge simple-zonal-private-local":        this variable.
Step #27 - "converge simple-zonal-private-local":        
Step #27 - "converge simple-zonal-private-local":        Error: No value for required variable
Step #27 - "converge simple-zonal-private-local":        
Step #27 - "converge simple-zonal-private-local":          on variables.tf line 33:
Step #27 - "converge simple-zonal-private-local":          33: variable "compute_engine_service_accounts" {
Step #27 - "converge simple-zonal-private-local":        
Step #27 - "converge simple-zonal-private-local":        The root module input variable "compute_engine_service_accounts" is not set,
Step #27 - "converge simple-zonal-private-local":        and has no default value. Use a -var or -var-file command line argument to
Step #27 - "converge simple-zonal-private-local":        provide a value for this variable.
Step #27 - "converge simple-zonal-private-local":        
Step #27 - "converge simple-zonal-private-local":        Error: No value for required variable
Step #27 - "converge simple-zonal-private-local":        
Step #27 - "converge simple-zonal-private-local":          on variables.tf line 38:
Step #27 - "converge simple-zonal-private-local":          38: variable "registry_project_ids" {
Step #27 - "converge simple-zonal-private-local":        
Step #27 - "converge simple-zonal-private-local":        The root module input variable "registry_project_ids" is not set, and has no
Step #27 - "converge simple-zonal-private-local":        default value. Use a -var or -var-file command line argument to provide a
Step #27 - "converge simple-zonal-private-local":        value for this variable.
Step #27 - "converge simple-zonal-private-local": >>>>>> Running the command `terraform apply -auto-approve -lock=true -lock-timeout=0s -input=false -no-color -parallelism=10 -refresh=true  ` failed due to a non-zero exit code of 1.
Step #27 - "converge simple-zonal-private-local": >>>>>> ------Exception-------
Step #27 - "converge simple-zonal-private-local": >>>>>> Class: Kitchen::ActionFailed
Step #27 - "converge simple-zonal-private-local": >>>>>> Message: 1 actions failed.
Step #27 - "converge simple-zonal-private-local": >>>>>>     Converge failed on instance <simple-zonal-private-local>.  Please see .kitchen/logs/simple-zonal-private-local.log for more details
Step #27 - "converge simple-zonal-private-local": >>>>>> ----------------------
Step #27 - "converge simple-zonal-private-local": >>>>>> Please see .kitchen/logs/kitchen.log for more details
Step #27 - "converge simple-zonal-private-local": >>>>>> Also try running `kitchen diagnose --all` for configuration
```

bharathkkb commented:
@Monkeyanator I noticed it when working on another PR. You can switch to using the registry source for `gke-project-asm`  https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/99e8f9b88eff8363eb7a06faff79e460c049ac35/test/setup/main.tf

bharathkkb commented:
Opened https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/1159 since it was affecting a few PRs

Monkeyanator commented:
Thanks for looking into it @bharathkkb, made the change from #1159 here but seems like we still run into issues about resources changing (hard to pin down where the step actually fails from the logs). Think the PR with just the fix is failing CI as well

bharathkkb commented:
@Monkeyanator seemed like the failure in https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/1159 was a flake. I have kicked it off again. 

Also could we add some kind of upgrade guide for users coming from the old module? Is there a possible upgrade path minimizing disruption?

Monkeyanator commented:
@bharathkkb the exposed options and implementation are different enough from the previous implementation where it'll be tough to support a migration path here. Is it possible to document to pin to the old version for this first iteration? Updated the README documenting this.

bharathkkb commented on modules/asm/README.md:
```diff
@@ -2,87 +2,37 @@
+1. Register target cluster to the servicemesh-enabled hub
 
-## Requirements
+## Migration
```
Lets create an upgrade guide for v20 and move this there.
https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/docs

Monkeyanator replied:
Done.

bharathkkb commented on modules/asm/main.tf:
```diff
@@ -14,95 +14,57 @@
-module "asm-services" {
-  source  = "terraform-google-modules/project-factory/google//modules/project_services"
-  version = "~> 11.3"
+#resource "google_gke_hub_feature" "mesh" {
```
nit: lets track this in an issue rather than commented code

Monkeyanator replied:
Removed the commented code.

bharathkkb commented on modules/asm/scripts/create_cpr.sh:
```diff
@@ -0,0 +1,59 @@
+
+set -e
+
+if [ "$#" -lt 3 ]; then
```
Seemed like we expected 4 args?
```suggestion
if [ "$#" -lt 4 ]; then
```

Monkeyanator replied:
Fixed, thanks for catching.

bharathkkb commented on modules/asm/README.md:
````diff
@@ -2,87 +2,37 @@
 
-There is a [full example](../../examples/simple_zonal_with_asm) provided. Detailed usage example is as follows:
-
-```tf
````
We should continue to have a section with Terraform snippet showcasing minimal usage

Monkeyanator replied:
Added back in.

bharathkkb commented:
> the exposed options and implementation are different enough from the previous implementation where it'll be tough to support a migration path here.

Makes sense, I dropped a comment above to move that section to the upgrade guide.

gottfrois commented:
Super excited that this PR has been merged. Do you think it will solve this issue? https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/issues/1114#issuecomment-1057249680

Monkeyanator commented:
> Super excited that this PR has been merged. Do you think it will solve this issue? #1114

Yes, the segfault probably comes from somewhere in the `install_asm` script which we completely ditched in the new module

elebioda commented:
Why was asm_version removed?
